gpflow/mean_functions.py

```diff
@@ -152,6 +154,46 @@ def __call__(self, X: TensorType) -> tf.Tensor:
         return tf.zeros(output_shape, dtype=X.dtype)
 
 
+class Polynomial(MeanFunction):
+    """
+    A generic polynomial mean function.
+    """
+
+    @check_shapes("w: [broadcast output_dim, broadcast n_terms]")
+    def __init__(
+        self, degree: int, input_dim: int = 1, output_dim: int = 1, w: Optional[TensorType] = None
+    ) -> None:
+        """
+        :param degree: The degree of the polynomial.
+        :param input_dim: Number of inputs / variables this polynomial is defined over.
+        :param output_dim: Number of outputs / polynomials.
+        :param w: Initial weights of the terms of the polynomial.
+        """
+        powers = cs(tuple(self._compute_powers(degree, input_dim)), "[n_terms, input_dim]")
+        if w is None:
+            w = cs([1.0] + (len(powers) - 1) * [0.0], "[n_terms]")
+        w_shape = (output_dim, len(powers))
+        self.powers = tf.constant(powers, dtype=default_float())
+        self.w = Parameter(tf.broadcast_to(w, w_shape))
+
+    def _compute_powers(self, degree: int, input_dim: int) -> Iterator[Tuple[int, ...]]:
+        if not input_dim:
+            yield ()
+            return
+        for i in range(degree + 1):
+            for inner in self._compute_powers(degree - i, input_dim - 1):
+                yield (i,) + inner
+
+    @check_shapes(
+        "X: [batch..., input_dim]",
+        "return: [batch..., output_dim]",
+    )
+    def __call__(self, X: TensorType) -> tf.Tensor:
```

Comment:
Maybe we can reply on `tf.math.polyval` to evaluate the polynomial: https://www.tensorflow.org/api_docs/python/tf/math/polyval. I believe uses Horner's method, a slightly more stable alternative to evaluate polynomials.

Reply:
> Maybe we can reply on `tf.math.polyval` to evaluate the polynomial: https://www.tensorflow.org/api_docs/python/tf/math/polyval. I believe uses Horner's method, a slightly more stable alternative to evaluate polynomials.

Yeah, I was looking at whether there was a library function for this - I think `tf.math.polyval` only applies to the single-variate case.